### DISCUSSION TITLE
How to capture diagonostic through python api or http endpoint instead of html

### DISCUSSION THREAD

**Discussion by zqbake** [dask/distributed#4669] [Q&A] [answered]
In which object I can get this information or if it's possible to write plugins to access these data collected by dask.distributed.

**Comment by jrbourbeau** [comment]
Thanks for asking @zhqu1148980644. Could you be more specific about what information you'd like to capture?

**Reply by zqbake** [reply]
@jrbourbeau  all information on the dashboard, like cpu, load balance per worker, etc, is there any programmatic way to fetch them?

**Reply by zqbake** [reply]
@jrbourbeau Hi, do you have any ideas, sincerely appreciate your help.

**Comment by fjetter** [comment, marked as answer]
## Accessing through HTTP (prometheus metrics)

There exists a `/metrics` endpoint which is used to expose some of this information as [prometheus](https://prometheus.io/) metrics. The exposed metrics for scheduler and workers differ slightly.
Unfortunately, this is not, yet, part of the official documentation but if you are looking for the code, see [here](https://github.com/dask/distributed/blob/main/distributed/http/scheduler/prometheus/__init__.py) and [here](https://github.com/dask/distributed/blob/main/distributed/http/worker/prometheus.py).

There are also some JSON endpoints available, e.g. `/json/counts.json`

All available endpoints are listed [here](https://distributed.dask.org/en/latest/http_services.html)

## Python API

Afaik, there is not one dedicated API to extract statistics. Depending on what you are looking for the distributed [Client](https://distributed.dask.org/en/latest/api.html) might offer some, if not all information. For instance, there is the [Client.scheduler_info](https://distributed.dask.org/en/latest/api.html#distributed.Client.scheduler_info).

There is of course always the possibility to write your own [plugin](https://distributed.dask.org/en/latest/plugins.html) to extract all the custom information you require.

**Reply by zqbake** [reply]
@fjetter  Thanks, I choose to write my own monitoring plugin. By the way,  can I use another thread to collect stats and then send it back to the calling side of `client.submit`? how can I achieve this, can I attach it to the submit-returned future?

**Reply by fjetter** [reply]
IIUC, you want to have custom, task/computation specific metadata attached to the futures to pass along your metrics? Afaik, there is no mechanism for this available.

**Reply by zqbake** [reply]
@fjetter How about wrapping my computation function/task into a function that automatically starts a new thread for collecting resource usages and sends it back as return value? if it's doable. And are there any better ways to achieve this?

**Reply by fjetter** [reply]
If you want to instrument _your_ code this should be a possibility. If you want to get metrics from `dask` itself you won't have access to the worker instance this way and there won't be a way to get this information, I am afraid.

**Comment by mrocklin** [comment]
There are some JSON routes.  Unfortunately these aren't well documented.
Rather than client.submit you might want to look at client.run.

**Reply by zqbake** [reply]
I'd like to collect diagnostics per job.
